### PR TITLE
[hir] Convert HIR to JS

### PR DESCRIPTION
**Pull request by meganyin13**
Convert HIR AST to valid JS

### CODE REVIEW

**Comment by SamChou19815** [comment]
I think you need to at least implement these functions instead of directly calling `JSON.stringify`:

- `highIRExpressionToString`
- `highIRStatementToString`
- `highIRFunctionToString`

Each functions prints the high-ir AST node in equivalent JS code.

**Comment by meganyin13** [comment]
Happy reviewing 🎉  there are some _easter eggs_ in these test cases 👀

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 1]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
...
+          },
+        ],
+        s2: [
+          {
+            __type__: 'HighIRReturnStatement',
+          },
+        ],
+      })
+    )
+  ).toBe(`if (5 == 5) {} else {}`);
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:125]: It looks like a bug. The return statement seems to be ignored by toString

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 2]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
...
+      parameters: ['d', 't', 'i'],
+      hasReturn: true,
+      body: [
+        HIR_LET({
+          name: 'b',
+          assignedExpression: HIR_INT(BigInt(1857)),
+        }),
+      ],
+    })
+  ).toBe(`const baz = (d, t, i) => {let b = 1857;; return;}`);
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:168]: Could we get rid of the double ;;

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 3]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
...
+  ).toBe(`if (5 == 5) {} else {}`);
+  expect(
+    highIRStatementToString(
+      HIR_FUNCTION_CALL({
+        functionArguments: [],
+        functionExpression: HIR_NAME('func'),
+        returnCollector: 'val',
+      })
+    )
+  ).toBe('let val = func();');
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:134]: `let` should be `var`. See https://samlang.developersam.com/docs/intermediate-representation#hir-statements

File: samlang-core/src/services/source-processor.ts [hunk 4]
```diff
@@ -52,3 +65,78 @@ export const lowerSourcesToAssemblyPrograms = (
...
+      )}(${functionArguments.map((arg) => highIRExpressionToString(arg)).join(', ')});`;
+    }
+    case 'HighIRLetDefinitionStatement': {
+      const { name, assignedExpression } = highIRStatement as HighIRLetDefinitionStatement;
+      return `let ${name} = ${highIRExpressionToString(assignedExpression)};`;
+    }
+    case 'HighIRReturnStatement':
+      return highIRStatement.expression
+        ? `return ${highIRExpressionToString(highIRStatement.expression)};`
+        : '';
```
Comment (SamChou19815) [inline samlang-core/src/services/source-processor.ts:95]: You still need `return;` here.

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 5]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
...
+  expect(
+    highIRStatementToString(
+      HIR_LET({
+        name: 'foo',
+        assignedExpression: HIR_INT(BigInt(19815)),
+      })
+    )
+  ).toBe(`let foo = 19815;`);
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:142]: 👀

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 6]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
...
+      })
+    )
+  ).toBe(`st = [0, 'bar', 13];`);
+});
+
+it('HIR function to JS string test', () => {
+  expect(
+    highIRFunctionToString({
+      name: 'baz',
+      parameters: ['d', 't', 'i'],
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:159]: 🚀

File: samlang-core/src/services/__tests__/source-processor.test.ts [hunk 7]
```diff
@@ -64,3 +86,108 @@ GLOBAL_STRING_1:
     .text
 `);
 });
+
+it('compile hello world to JS integration test', () => {
+  const moduleReference = new ModuleReference(['Test']);
+  const sourceCode = `
+    class Main {
+        function main(): unit = println('Hello '::'World!')
```
Comment (SamChou19815) [inline samlang-core/src/services/__tests__/source-processor.test.ts:94]: There is an syntax error here: samlang strings use double-quote. The syntax error causes the AST node to be dropped, that's why you see some garbage in JS output.
Classic garbage-in-garbage-out example :)

File: samlang-core/src/services/source-processor.ts [hunk 8]
```diff
@@ -52,3 +65,78 @@ export const lowerSourcesToAssemblyPrograms = (
...
+};
+
+export const highIRExpressionToString = (highIRExpression: HighIRExpression): string => {
+  switch (highIRExpression.__type__) {
+    case 'HighIRIntLiteralExpression':
+      return `${highIRExpression.value}`;
+    case 'HighIRStringLiteralExpression':
+      return `'${highIRExpression.value}'`;
+    case 'HighIRIndexAccessExpression': {
+      const { expression, index } = highIRExpression as HighIRIndexAccessExpression;
```
Comment (SamChou19815) [inline samlang-core/src/services/source-processor.ts:122]: Why do we need a cast here?

**Comment by SamChou19815** [comment]
And my opinion: It should probably be moved to the printer directory. Then we can later decide whether we want to adopt the prettier framework for JS output.

File: samlang-core/src/services/__tests__/source-processor.test.ts (hunk 3 above)
Comment (meganyin13) [reply samlang-core/src/services/__tests__/source-processor.test.ts:134]: but `var` is bad :/

Comment (SamChou19815) [reply samlang-core/src/services/__tests__/source-processor.test.ts:134]: It's bad, but it's necessary. It's the way how HIR lowering works result in this decision:

Consider this if-expression `if (b) then { val _ = println("a");  "a" } else { val _ = println("b");  "b" }`

It is translated into this HIR (written is JS syntax):

```js
if (b) {
 println("a");
 var _temp0 = "a";
} else {
 println("b");
 var _temp0 = "b";
}
_temp0 // as result of this expression
```

You can see that we have to abuse `var`'s function scope capability.

File: samlang-core/src/printer/printer-js.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+export const highIRExpressionToString = (highIRExpression: HighIRExpression): string => {
+  switch (highIRExpression.__type__) {
+    case 'HighIRIntLiteralExpression':
+      return `${highIRExpression.value}`;
+    case 'HighIRStringLiteralExpression':
+      return `'${highIRExpression.value}'`;
+    case 'HighIRIndexAccessExpression': {
+      const { expression, index } = highIRExpression;
+      return `${highIRExpressionToString(expression)}[${index}]`;
```
Comment (SamChou19815) [inline samlang-core/src/printer/printer-js.ts:59]: To be safe, we need to add parenthesis around the expression. Same for the binary expression case below.

File: samlang-core/src/printer/__tests__/printer-js.test.ts [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  const moduleReference = new ModuleReference(['Test']);
+  const sourceCode = `
+    class Main {
+        function main(): unit = println("Hello "::"World!")
+    }
+    `;
+  const { checkedSources } = checkSources([[moduleReference, sourceCode]]);
+  const hirSources = compileSamlangSourcesToHighIRSources(checkedSources);
+  expect(highIRSourcesToJSString(hirSources)).toBe(
+    `{const _module_Test_class_Main_function_main = () => {var _t0 = _builtin_stringConcat('Hello ', 'World!');;var _t1 = _builtin_println(_t0); };}`
```
Comment (SamChou19815) [inline samlang-core/src/printer/__tests__/printer-js.test.ts:34]: We probably don't need the `{` and `}` around the code

File: samlang-core/src/printer/printer-js.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+      const { e1, e2, operator } = highIRExpression;
+      return `${highIRExpressionToString(e1)} ${operator} ${highIRExpressionToString(e2)}`;
+    }
+  }
+};
+
+export const highIRSourcesToJSString = (sources: Sources<HighIRModule>): string => {
+  let finalStr = '';
+  sources.forEach((module) => {
+    finalStr += `{${module.functions.map((f) => highIRFunctionToString(f)).join(';')}}`;
```
Comment (SamChou19815) [inline samlang-core/src/printer/printer-js.ts:75]: Consider join by `;\n` so it's more readable.

File: samlang-core/src/printer/__tests__/printer-js.test.ts [hunk 12]
```diff
@@ -123,5 +123,5 @@ it('HIR expression to JS string test', () => {
         e2: HIR_INT(BigInt(7)),
       })
     )
-  ).toBe('7 != 7');
+  ).toBe('{7 != 7}');
```
Comment (SamChou19815) [inline samlang-core/src/printer/__tests__/printer-js.test.ts:126]: well, I think it should be `(7) != (7) instead of `{7 != 7}`

Comment (meganyin13) [reply samlang-core/src/printer/__tests__/printer-js.test.ts:126]: ohhh you meant `(` and `)`

Comment (meganyin13) [reply samlang-core/src/printer/__tests__/printer-js.test.ts:126]: wait I think the binary case should be `(7 != 7)` unless we want `(` and `)` around int literals as well?

Comment (SamChou19815) [reply samlang-core/src/printer/__tests__/printer-js.test.ts:126]: Good point. I think we only need to add parenthesis around the entire binary expression.

Comment (meganyin13) [reply samlang-core/src/printer/__tests__/printer-js.test.ts:126]: And for the `(samlang)[3]` case the parenthesis would be around the `HighIRVariableExpression` and `HighIRNameExpression`?

Comment (SamChou19815) [reply samlang-core/src/printer/__tests__/printer-js.test.ts:126]: We probably don't need to do anything special with the index expression. It has a very high precedence level in JS, so we don't need parenthesis to disambiguate.

**Pull request merged**